driver-core/src/main/com/mongodb/connection/CompositeByteBuf.java

```diff
@@ -0,0 +1,343 @@
+    }
+
+    @Override
+    public ByteBuf get(final int index, final byte[] bytes, final int offset, final int length) {
+        checkDstIndex(index, length, offset, bytes.length);
+
+        int i = findComponentIndex(index);
+        int curIndex = index;
+        int curOffset = offset;
+        int curLength = length;
+        while (curLength > 0) {
+            Component c = components.get(i);
+            int localLength = Math.min(curLength, c.buffer.capacity() - (curIndex - c.offset));
+            c.buffer.get(curIndex - c.offset, bytes, curOffset, localLength);
+            curIndex += localLength;
+            curOffset += localLength;
+            curLength -= localLength;
+            i++;
+        }
+
+        return this;
+    }
+
+    @Override
+    public long getLong() {
+        position += 8;
+        return getLong(position - 8);
+    }
+
+    @Override
+    public long getLong(final int index) {
+        checkIndex(index, 8);
+        Component component = findComponent(index);
+        if (index + 8 <= component.endOffset) {
+            return component.buffer.getLong(index - component.offset);
+        } else {
+            return getInt(index) & 0xFFFFFFFFL | (getInt(index + 4) & 0xFFFFFFFFL) << 32;
+        }
+    }
+
+    @Override
+    public double getDouble() {
+        position += 8;
+        return getDouble(position - 8);
+    }
+
+    @Override
+    public double getDouble(final int index) {
+        return Double.longBitsToDouble(getLong(index));
+    }
+
+    @Override
+    public int getInt() {
+        position += 4;
+        return getInt(position - 4);
+    }
+
+    @Override
+    public int getInt(final int index) {
+        checkIndex(index, 4);
+        Component component = findComponent(index);
+        if (index + 4 <= component.endOffset) {
+            return component.buffer.getInt(index - component.offset);
+        } else {
+            return getShort(index) & 0xFFFF | (getShort(index + 2) & 0xFFFF) << 16;
+        }
+    }
+
+    private int getShort(final int index) {
+        checkIndex(index, 2);
+        return (short) (get(index) & 0xff | (get(index + 1) & 0xff) << 8);
+    }
+
+    @Override
+    public byte[] array() {
+        throw new UnsupportedOperationException("Not implemented yet!");
+    }
+
+    @Override
+    public ByteBuf limit(final int newLimit) {
+        if (newLimit < 0 || newLimit > capacity()) {
+            throw new IndexOutOfBoundsException(format("%d is out of bounds", newLimit));
+        }
+        this.limit = newLimit;
+        return this;
+    }
+
+    @Override
+    public ByteBuf put(final int index, final byte b) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ByteBuf put(final byte[] src, final int offset, final int length) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ByteBuf put(final byte b) {
+        throw new UnsupportedOperationException();
```

Comment:
As this is an internal class only used for reading from the byte buffer, these should never be called.